omegaconf/_utils.py

```diff
@@ -378,22 +383,13 @@ def ret(
     if value == "???":
         return ret(ValueKind.MANDATORY_MISSING)
 
-    if not isinstance(value, str):
-        return ret(ValueKind.VALUE)
-
-    match_list = list(re.finditer(key_prefix + legal_characters, value))
-    if len(match_list) == 0:
+    if not isinstance(value, str) or "${" not in value:
         return ret(ValueKind.VALUE)
 
-    if len(match_list) == 1 and value == match_list[0].group(0):
-        return ret(ValueKind.INTERPOLATION)
-    else:
-        return ret(ValueKind.STR_INTERPOLATION)
-
+    if return_parse_tree:
+        parse_tree = grammar_parser.parse(value)
 
-def is_bool(st: str) -> bool:
-    st = str.lower(st)
-    return st == "true" or st == "false"
+    return ret(ValueKind.INTERPOLATION)
```

Comment:
I assume your question relates to more than just the few lines highlighted in your comment. There are a couple of things going on there:

1. Removal of `ValueKind.STR_INTERPOLATION`: as far as I could tell there was no need anymore to differentiate it from `ValueKind.INTERPOLATION`, because now string interpolations can be parsed with the same code as "simple" interpolations.

2. New logic to decide what is an interpolation vs. what isn't. The old code was using this regex matching logic that doesn't really make sense anymore with the new grammar. For instance with that old code, if you had an interpolation using some invalid characters, it would behave as if it was a regular string, while what we actually want (with the new grammar version) is to raise an exception during parsing. So my reasoning for using the presence of `${` in the string is that (1) if it's not there we are 100% sure there is no interpolation, and (2) if it's there then we should consider that it's an interpolation by default, and let the grammar parser + visitor validate later that it is well formed. Early on I actually tried to actually parse the expression here in order to make that decision (even with `return_parse_tree` set to False), however it incurred a heavy overhead because there are frequent calls to `get_value_kind(..., return_parse_tree=False)`. On the other hand, checking for the presence of `${` in the string is much more efficient. I added a new comment in 5bbd76c to make it clearer.

Reply:
1. Okay, if we can eliminate the need to special handling of STR_INTERPOLATION its' a win.
2. the question is more basic than this:
This function checks of a string represents a boolean value.
```
def is_bool(st: str) -> bool:
    st = str.lower(st)
    return st == "true" or st == "false"
```

After your fix, I have no idea what it does and why it's called is_bool.

Reply:
> After your fix, I have no idea what it does and why it's called is_bool.

Oh, actually that function has been removed in this diff, probably easier to see in a side-by-side diff: https://github.com/omry/omegaconf/pull/321/files/76ebe1abf9aa383d5fe97650b383bbd27cfc7f65..12424eebba55185e5e4710dcd302041f50ae8401#diff-3dae9501bcb701903e5b292b8e1e0637

Reply:
And thinking about it, it's probably a mistake on my side when I rebased all my changes to split them into individual commits, the removal of `is_bool()` should have been in the commit where I removed `decode_primitive()` (b88a270c3178603f84b3aaad9bcbb6ad59816681)

Reply:
Okay, once I am done with the step by step review I will make a final review on the combined diff.
the change I am seeing here is changing the meaning of the function which is why I brought it up.